db/integrity/commitment.go

```diff
@@ -0,0 +1,428 @@
+						"branchKey", hex.EncodeToString(branchKey),
+						"addr", common.BytesToAddress(key[:length.Addr]),
+						"kv", fileName,
+					)
+				}
+				counts.accNonDerefCount++
+				return key, nil // not a referenced key, nothing to check
+			}
+			counts.accDerefCount++
+			offset, err := checkOffsetDeref(key, uint64(accReader.Size()))
+			if err != nil {
+				err = fmt.Errorf("account reference key %x issue for branch %x in %s: %w", key, branchKey, fileName, err)
+				if failFast {
+					return nil, err
+				}
+				logger.Warn(err.Error())
+				return key, nil
+			}
+			accReader.Reset(offset)
+			plainKey, _ := accReader.Next(plainKeyBuf[:0])
+			if len(plainKey) != length.Addr {
+				err = fmt.Errorf("account reference key %x has invalid plainKey for branch %x in %s", key, branchKey, fileName)
+				if failFast {
+					return nil, err
+				}
+				logger.Warn(err.Error())
+				integrityErr = fmt.Errorf("%w: %w", ErrIntegrity, err)
+				return key, nil
+			}
+			if logger.Enabled(ctx, log.LvlTrace) {
+				logger.Trace(
+					"dereferenced account key",
+					"branchKey", hex.EncodeToString(branchKey),
+					"key", hex.EncodeToString(key),
+					"offset", offset,
+					"addr", common.BytesToAddress(plainKey),
+					"kv", fileName,
+				)
+			}
+			return plainKey, nil
+		})
+		if err != nil {
+			return derefCounts{}, err
+		}
+		err = newBranchData.Validate(branchKey)
+		if err != nil {
+			err = fmt.Errorf("branch data validation failure for branch key %x in %s: %w", branchKey, fileName, err)
+			if failFast {
+				return derefCounts{}, err
+			}
+			logger.Warn(err.Error())
+			integrityErr = fmt.Errorf("%w: %w", ErrIntegrity, err)
+		}
+	}
+	logger.Info(
+		"checked commitment kv dereference in",
+		"dur", time.Since(start),
+		"keys", counts.branchKeyCount,
+		"accDerefs", counts.accDerefCount,
+		"accNonDerefs", counts.accNonDerefCount,
+		"storageDerefs", counts.storageDerefCount,
+		"storageNonDerefs", counts.storageNonDerefCount,
+		"kv", fileName,
+	)
+	return counts, integrityErr
+}
+
+func deriveReaderForOtherDomain(baseFile string, oldDomain, newDomain kv.Domain) (*seg.Reader, func(), error) {
+	fileVersionMask, err := version.ReplaceVersionWithMask(baseFile)
+	if err != nil {
+		return nil, nil, err
+	}
+	fileVersionMask = strings.Replace(fileVersionMask, oldDomain.String(), newDomain.String(), 1)
+	newFile, _, ok, err := version.FindFilesWithVersionsByPattern(fileVersionMask)
+	if err != nil {
+		return nil, nil, err
+	}
+	if !ok {
+		return nil, nil, fmt.Errorf("could not derive reader for other domain due to file not found: %s,%s->%s", baseFile, oldDomain, newDomain)
+	}
+	decomp, err := seg.NewDecompressor(newFile)
+	if err != nil {
+		return nil, nil, err
+	}
+	compression := statecfg.Schema.GetDomainCfg(newDomain).Compression
+	return seg.NewReader(decomp.MakeGetter(), compression), decomp.Close, nil
+}
+
+func checkOffsetDeref(key []byte, end uint64) (uint64, error) {
+	offset, n := binary.Uvarint(key)
```

Comment:
can reuse logic from `decodeShorterKey` (better to rename to something `ref/deref/offset` related and export i think?) in db/state/domain_committed` to at least keep a link on this code